working/0546-patterns/patterns-feature-specification.md

```diff
@@ -1682,130 +1887,138 @@ At runtime, a pattern is matched against a value. This determines whether or not
+        fails.
 
-    2.  For each `mapBinderEntry` or `mapMatcherEntry`:
+    2.  Otherwise, match the right subpattern against `v` and succeed if it
+        matches.
 
-        1.  Evaluate key `expression` to `key` and call `containsKey()` on
-            the value. If this returns `false`, the map does not match.
+*   **Relational**:
 
-        3.  Otherwise, evaluate `v[key]` and match the resulting value against
-            this entry's value subpattern. If it does not match, the map does
-            not match.
+    1.  Call the corresponding operator from `p` on `v` passing the constant
+        value from `p` as the argument.
 
-    3.  If all entries match, the map matches.
+    2.  If the method returns `true`, the match succeeds, otherwise it fails.
 
-    *Note that, unlike with lists, a matched map may have additional entries
-    that are not checked by the pattern.*
+*   **Null-check**:
 
-*   **Record matcher or binder**:
+    1.  If `v` is null then the match fails.
 
-    1.  If `p` has positional fields and `v` has type `dynamic`, then throw a
-        runtime exception if `v` is not an instance of the inferred record type
-        of `p`. *Edge case: Record patterns with positional fields can only
-        match record objects. Normally we statically ensure that a record
-        pattern with positional fields can only match a value known to be a
-        record type. But `dynamic` evades that check, so check here. We only do
-        this if the pattern has positional fields in order to allow record
-        patterns with only named fields to be used to call arbitrary getters on
-        values of type `dynamic`.*
+    2.  Otherwise, match the inner pattern against `v`.
 
-    2.  For each field `f` in `p`, in source order:
+*   **Null-assert**:
 
-        1.  If `f` is positional, then destructure the corresponding positional
-            field from record `v` to get result `r`.
+    1.  If `v` is null then throw a runtime exception. *Note that we throw even
+        if this appears in a refutable context. The intent of this pattern is to
+        assert that a value *must* not be null.*
 
-        2.  Otherwise (`f` is named), call the getter with the same name as `f`
-            on `v` to get result `r`. *If `v` has type `dynamic`, this getter
-            call may throw a NoSuchMethodError, which we allow to propagate
-            instead of treating that as a match failure.*
+    2.  Otherwise, match the inner pattern against `v`.
 
-        3.  Match the subpattern of `f` against `r`. If the match fails, the
-            record match fails. If `f` is a named field using the shorthand
-            syntax that that infers an implicit variable subpattern from the
-            field's name, match `r` against that inferred variable subpattern.
+*   **Literal** or **constant**: The pattern matches if `o == v` evaluates to
+    `true` where `o` is the pattern's value.
```

Comment:
It's probably a good call on using `o == v`. The pattern decides whether the value matches, not the other way around.

Using `==` is still dangerous, since it calls user code and can act erratically.
It does mean that `c` is the same as `== c`, which is convenient.
It does mean there is no easy way to check for identity.
Guess you can always do:
```dart
class Ider {
  final Object? value;
  Ider(this.value);
  int get hashCode => identityHasHCode(value);
  bool operator==(Object other) => identical(value, other);
}
```
and use that in your patterns. 
(Which is why `v == o` might also be a valid choice, because then the pattern author can fiddle with equality.)

Reply:
Yeah, I think in practice this will almost always be used with types that have entirely boring equality behavior, so it doesn't matter too much what we do, as long as it's specified precisely.